[paplayer] fix codec deinit for audio decoder add-on (was never called)

AlwinEsch opened the pull request:
## Description

Fix of never called decoder deinit
## Motivation and Context

Has found during my add-on system rework that on audio decoder add-on the `DeInit` was never called and a mem leak with this come.

This add the use of `ICodec::DeInit` call to the `CAudioDecoder::Destroy()` function.

On the other class (`VideoPlayerCodec`) who use also the `ICodec` is it done from his destructor.
## Types of change

<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply -->

<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

AlwinEsch commented on xbmc/cores/paplayer/AudioDecoder.cpp:
```diff
@@ -55,6 +55,8 @@ void CAudioDecoder::Destroy()
+    m_codec->DeInit();
   m_pcmBuffer.Destroy();
 
   if ( m_codec )
```
Maybe another way can be to add it on `virtual ~ICodec() {};`???

FernetMenta commented:
thanks bob for pinging the right people :)

FernetMenta commented on xbmc/cores/paplayer/AudioDecoder.cpp:
```diff
@@ -58,7 +58,10 @@ void CAudioDecoder::Destroy()
 
   if ( m_codec )
+  {
+    m_codec->DeInit();
```
I don't see what is here. Now you call Deinit before the destructor.
This change violates the principle of encapsulation. Why do you want to call DeInit from the outside?

FernetMenta replied:
What does this entire PR change in behaviour?

AlwinEsch replied:
An alternative way can be to remove the:
```cpp
  // DeInit()
  // Should just cleanup anything as necessary.  No need to free buffers here if they
  // are allocated and destroyed in the destructor.
  virtual void DeInit()=0;
```
complete and to do this on destructors from `ADDON::CAudioDecoder` and like before on `VideoPlayerCodec`. What do you think about them?

FernetMenta replied:
please answer my question first

AlwinEsch replied:
It change in behaviour nearly nothing, the only new thing is that a function who was not called on audio decoder add-on's before is called now.

Not sure if it was never called or it was lost on a update of PAPlayer.

AlwinEsch replied:
Since the `DeInit` is set in` ICodec` I am assuming that it will be called from the outside and not in the child class itself uses the `ICodec`.

FernetMenta replied:
ah ok, we are talking about an addon issue here. I would drop DeInit from the interface ICodec and make the addon CAudioDecoder behave like VideoPlayerCodec. Do deinit in the destructor: https://github.com/xbmc/xbmc/blob/master/xbmc/addons/AudioDecoder.h#L59

AlwinEsch committed:
[paplayer] fix codec deinit (was never called)

It calls DeInit now on ADDON::CAudioDecoder like on VideoPlayerCodec.
Also becomes the DeInit removed from ICodec since it is handled on the
classes themselves.

AlwinEsch commented:
Is updated and during test the DeInit on add-on itself was called good.

jenkins build this please

FernetMenta commented:
+1
